File: app/src/main/java/ua/hospes/rtm/ui/cars/CarsPresenter.java
package ua.hospes.rtm.ui.cars;

import javax.inject.Inject;

import ua.hospes.rtm.core.ui.BasePresenter;
import ua.hospes.rtm.domain.cars.CarsInteractor;
import ua.hospes.rtm.utils.RxUtils;

/**
 * @author Andrew Khloponin
 */
class CarsPresenter extends BasePresenter<CarsContract.View> {
    private final CarsInteractor interactor;

    @Inject
    CarsPresenter(CarsInteractor interactor) {
        this.interactor = interactor;
    }

    @Override
    public void attachView(CarsContract.View view) {
        super.attachView(view);

        RxUtils.manage(this, interactor.listen()
                .compose(RxUtils.applySchedulers())
                .subscribe(list -> getView().updateCars(list), Throwable::printStackTrace));
    }

    @Override
    public void detachView() {
        super.detachView();
        RxUtils.unsubscribe(this);
    }


    public void removeAll() {
        RxUtils.manage(this, interactor.removeAll()
                .compose(RxUtils.applySchedulersSingle())
                .subscribe(result -> {}, Throwable::printStackTrace));
    }
}

File: app/src/main/java/ua/hospes/rtm/ui/cars/CarsPresenter.kt
package ua.hospes.rtm.ui.cars

import androidx.lifecycle.Lifecycle
import dagger.hilt.android.scopes.ActivityRetainedScoped
import dagger.hilt.android.scopes.ActivityScoped
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import ua.hospes.rtm.core.Presenter
import ua.hospes.rtm.data.CarsRepository
import javax.inject.Inject

class CarsPresenter @Inject constructor(
        private val repo: CarsRepository
) : Presenter<CarsContract.View>() {

    override fun attachView(view: CarsContract.View?, lc: Lifecycle) {
        super.attachView(view, lc)

        launch { repo.listen().collect { view?.onData(it) } }
    }

    override fun onError(throwable: Throwable) = view?.onError(throwable) ?: Unit
    override fun onUnexpectedError(throwable: Throwable) = view?.onError(throwable) ?: Unit


    fun removeAll() = launch { repo.clear() }
}